### PR TITLE
[GEOS-9316] Support for vector custom dimensions

### PR DESCRIPTION
**Pull request by fernandor777**
This PR adds custom dimensions support for vector layers, as defined on GSIP 177.

https://github.com/geoserver/geoserver/wiki/GSIP-177

https://osgeo-org.atlassian.net/browse/GEOS-9316

## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [x] Confirm you have read the [contribution guidelines](https://github.com/geoserver/geoserver/blob/master/CONTRIBUTING.md) 
- [x] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [x] Make sure the first PR targets the master branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.

The following are required only for core and extension modules (they are welcomed, but not required, for community modules):
- [x] There is a ticket in Jira describing the issue/improvement/feature (a notable exemptions is, changes not visible to end users)
- [x] PR for bug fixes and small new features are presented as a single commit
- [x] Commit message must be in the form "[GEOS-XYZW] Title of the Jira ticket" (export to XML in Jira generates the message in this exact form)
- [x] New unit tests have been added covering the changes
- [x] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] This PR passes the [QA checks](https://docs.geoserver.org/latest/en/developer/qa-guide/index.html) (QA checks results will be reported by travis-ci after opening this PR)
- [x] Commits changing the UI, existing user workflows, or adding new functionality, need to include documentation updates

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or inapplicable.

### CODE REVIEW

**Comment by nmco** [comment]
I performed some manual tests using several layers based on the following PostgreSQL table:
```
CREATE TABLE points(
	feature_id serial PRIMARY KEY, 
	geometry geometry(POINT, 4326),
	attribute_1 timestamp,
	attribute_2 timestamp,
	attribute_3 text,
	attribute_4 text,
	attribute_5 integer,
	attribute_6 integer);

INSERT INTO points (
	geometry, 
	attribute_1, 
	attribute_2, 
	attribute_3, 
	attribute_4, 
	attribute_5, 
	attribute_6)
VALUES
   (ST_SetSRID(ST_MakePoint(0.0, 1.0), 4326), '2019-09-01 00:00:00', '2019-09-02 00:00:00', '#FF0000', '#0000FF', 01, 02),
   (ST_SetSRID(ST_MakePoint(1.5, 1.0), 4326), '2019-09-01 00:10:00', '2019-09-02 00:10:00', '#00FF00', '#0000FF', 03, 04),
   (ST_SetSRID(ST_MakePoint(1.0, 0.0), 4326), '2019-09-01 00:20:00', '2019-09-02 00:20:00', '#0000FF', '#0000FF', 05, 06),
   (ST_SetSRID(ST_MakePoint(2.5, 1.0), 4326), '2019-09-01 00:30:00', '2019-09-02 00:30:00', '#FF0000', '#00FF00', 07, 08),
   (ST_SetSRID(ST_MakePoint(0.0, 1.5), 4326), '2019-09-01 00:40:00', '2019-09-02 00:40:00', '#0000FF', '#00FF00', 09, 10),
   (ST_SetSRID(ST_MakePoint(2.0, 2.0), 4326), '2019-09-01 00:50:00', '2019-09-02 00:50:00', '#00FF00', '#FF0000', 11, 12),
   (ST_SetSRID(ST_MakePoint(0.0, 6.0), 4326), '2019-09-01 00:00:00', '2019-09-02 00:00:00', '#00FF00', '#0000FF', 01, 02),
   (ST_SetSRID(ST_MakePoint(1.5, 5.0), 4326), '2019-09-01 00:10:00', '2019-09-02 00:10:00', '#FF0000', '#0000FF', 03, 04),
   (ST_SetSRID(ST_MakePoint(1.0, 4.0), 4326), '2019-09-01 00:20:00', '2019-09-02 00:20:00', '#00FF00', '#0000FF', 05, 06),
   (ST_SetSRID(ST_MakePoint(1.0, 5.0), 4326), '2019-09-01 00:30:00', '2019-09-02 00:30:00', '#0000FF', '#00FF00', 07, 08),
   (ST_SetSRID(ST_MakePoint(0.0, 5.5), 4326), '2019-09-01 00:40:00', '2019-09-02 00:40:00', '#FF0000', '#00FF00', 09, 10),
   (ST_SetSRID(ST_MakePoint(2.0, 6.0), 4326), '2019-09-01 00:50:00', '2019-09-02 00:50:00', '#0000FF', '#FF0000', 11, 12);
```

**Issue 1**

To reproduce this issue, create the layer adding a custom dimension, for example:
image

Save the layer, then open it again a disable the custom dimension, when saving the layer an exception will be throw:

image

**Issue 2**

When creating a custom dimension, if we try to save without providing a mandatory value, like the dimension attribute:

image
 
The operation will not succeed, but no exception or message will be displayed. My expectation was to see a message (like for time and elevation) saying that a mandatory value was not provided.

**Issue 3**

When adding a new custom dimension there is not way back, let's say we start adding a new dimension, then we decide we don't need it, clicking in remove  will not do anything (no error, no message):

image

Disabling the dimension and clicking on remove it will throw the following exception:

```
(...)
Caused by: java.lang.NullPointerException
	at org.geoserver.web.data.resource.VectorCustomDimensionsPanel$VectorDimensionModel.setObject(VectorCustomDimensionsPanel.java:212)
	at org.geoserver.web.data.resource.VectorCustomDimensionsPanel$VectorDimensionModel.setObject(VectorCustomDimensionsPanel.java:181)
	at org.apache.wicket.Component.setDefaultModelObject(Component.java:3122)
	at org.apache.wicket.markup.html.form.FormComponent.setModelObject(FormComponent.java:1579)
	at org.apache.wicket.markup.html.form.FormComponent.updateModel(FormComponent.java:1097)
(...)
```
**Issue 4**

Selecting ``Use the reference value`` as default value does not activate the input to insert the reference value, this works as expected for elevation and time dimensions:

image

**Issue 5**

When selecting an end-value we have the possibility of selecting ``-`` not sure what's is propose.

** Issue 6**

It is possible to have the Time dimension enabled and create a custom dimension named time, this should probably be avoided 😋, same for elevation.

Trying to add a custom dimension twice (same name) will fail, but no error message is displayed, hence no visual feedback for the user.

Now I have a few questions, I configured the following dimensions:

image

image

image

image


Then I performed the following request:
```
http://localhost:8080/geoserver/wms?
SERVICE=WMS&
VERSION=1.1.1&
REQUEST=GetMap&
FORMAT=image/png&
TRANSPARENT=false&
LAYERS=points&
exceptions=application/vnd.ogc.se_inimage&
SRS=EPSG:4326&
STYLES=&
WIDTH=600&
HEIGHT=600&
BBOX=-1.812744140625,-0.362548828125,4.779052734375,6.229248046875&
Color%201=%23FF0000&
custom_time=2019-08-30T00:00:00.0Z/2019-09-02T23:59:59.999Z&
time=2019-06-30T00:00:00.0Z/2019-09-20T23:59:59.999Z
```

The SQL query send to the database is this one:

```
SELECT "feature_id", 
       "attribute_3", 
       Encode(St_astwkb(St_simplify(St_force2d("geometry"), 0.0087890625, true), 
              3), 
       'base64') AS "geometry" 
FROM   "public"."points" 
WHERE  ( "attribute_1" BETWEEN '2019-06-30 01:00:00.0' AND 
                               '2019-09-21 00:59:59.999' 
         AND "attribute_1" IS NOT NULL 
         AND "attribute_3" = '#FF0000' 
         AND "attribute_3" IS NOT NULL 
         AND "attribute_1" <= '2019-09-03 00:59:59.999' 
         AND "attribute_2" >= '2019-08-30 01:00:00.0' 
         AND "attribute_2" IS NOT NULL 
         AND "attribute_4" = '#0000FF' 
         AND "attribute_4" IS NOT NULL 
         AND "geometry" && St_geomfromtext( 
'POLYGON ((-1.856689453125 -0.406494140625, -1.856689453125 6.273193359375, 4.822998046875 6.273193359375, 4.822998046875 -0.406494140625, -1.856689453125 -0.406494140625))'
, 4326) ) 
```

I'm in doubt that in both situations (time and custom dimension) the attribute1 is treated as a time, I guess it is the case the difference here is that for the custom time we have and end time, could you confirm this behavior 😄?

**Comment by fernandor777** [comment]
Thanks for your review, UI issues 1 - 6 are now fixed.  About the point 6 doubt in fact the difference is because the endAttribute usage, both time and custom dimensions reuse the same Filter encoder.

**Comment by nmco** [comment]
I was doing one last round of tests and noticed that the reference value for custom dimension is not respected.

For example, configure a custom dimension:

image

Check the capabilities document, the reference will either the max or min value.

File: doc/en/user/source/data/webadmin/layers.rst [hunk 1]
```diff
@@ -366,3 +366,30 @@ Only for the "Reference value" strategy, and limited to times, it's also possibl
...
 
 .. note:: For more information on specifying times, please see the section on :ref:`wms_time`.
+
+Vector Custom Dimensions
+^^^^^^^^^^^^^^^^^^^^^^^^
+GeoServer also supports adding custom dimensions to vector layers, defining their names and configurations. 
+
+.. figure:: img/data_layers_dimension_editor_custom.png
+
+   Custom dimension enabled for a WMS layer
```
Comment (nmco) [inline doc/en/user/source/data/webadmin/layers.rst]: WMS layer -> vector layer

Comment (fernandor777) [reply doc/en/user/source/data/webadmin/layers.rst]: Fixed.

**Comment by fernandor777** [comment]
Default Strategy selection for custom dimensions fixed, thanks.

**Comment by nmco** [comment]
Great, thx @fernandor777 !

**Pull request merged**